src/core/server/server.ts

```diff
@@ -143,17 +145,28 @@ export class Server {
     );
   }
 
-  private registerCoreContext(coreSetup: InternalCoreSetup) {
-    coreSetup.http.registerRouteHandlerContext(coreId, 'core', async (context, req) => {
-      const adminClient = await coreSetup.elasticsearch.adminClient$.pipe(take(1)).toPromise();
-      const dataClient = await coreSetup.elasticsearch.dataClient$.pipe(take(1)).toPromise();
-      return {
-        elasticsearch: {
-          adminClient: adminClient.asScoped(req),
-          dataClient: dataClient.asScoped(req),
-        },
-      };
-    });
+  private registerCoreContext(coreSetup: {
+    http: HttpServiceSetup;
+    elasticsearch: ElasticsearchServiceSetup;
+    savedObjects: SavedObjectsServiceSetup;
+  }) {
+    coreSetup.http.registerRouteHandlerContext(
+      coreId,
+      'core',
+      async (context, req): Promise<RequestHandlerContext['core']> => {
+        const adminClient = await coreSetup.elasticsearch.adminClient$.pipe(take(1)).toPromise();
+        const dataClient = await coreSetup.elasticsearch.dataClient$.pipe(take(1)).toPromise();
+        return {
+          savedObjects: {
+            client: coreSetup.savedObjects.clientProvider.getClient(req),
```

Comment:
Do you know if there would be any advantages to using the same adminClient instance rather than having the clientProvider create its own? I'm not sure how the connection pooling mechanism works, but if they each have their own separate pools, that doesn't seem ideal.

Reply:
We use the same adminClient instance, we just subscribe to `coreSetup.elasticsearch.adminClient$` at a different place: https://github.com/elastic/kibana/blob/d896540a9daba324d5ac13ca0d82586f6084178d/src/core/server/saved_objects/saved_objects_service.ts#L128

We create a new instance of `ScopedClusterClient` but there is only ever two `ClusterClient`s one for data and one for admin.

Reply:
does it mean that `clientProvider` always uses the very first instance of the `adminClient` and ignores all updates? It's totally fine until we support config updates.

Reply:
yes, the outcome of https://github.com/elastic/kibana/pull/45796 might change how we handle this, but I will add a comment to the code since this is an important limitation to be aware of.